### PR TITLE
Prefer GNU init/destructor mechanism to be used for FIPS self tests

### PR DESCRIPTION
**Pull request by jon-oracle**
I've run across a problem on HP-UX where builds using gcc fail to run the init/cleanup functions. It seems that the compiler does not recognise the platform-specific #pragma init/fini statements.

By moving around the order of checks to ensure the __GNUC__ check is done first the build now works. If using the HP compiler the build/test also works fine.

I have checked that this mechanism works for gcc builds on AIX and Solaris also.

### CODE REVIEW

**Comment by t8m** [comment]
Please avoid the merge commits. Drop it and rebase the branch against current master.

File: providers/fips/self_test.c [hunk 1]
```diff
@@ -125,11 +133,6 @@ void _cleanup(void)
 # pragma init "init"
 # pragma fini "cleanup"
 
-#elif defined(__GNUC__)
-# undef DEP_INIT_ATTRIBUTE
-# undef DEP_FINI_ATTRIBUTE
-# define DEP_INIT_ATTRIBUTE static __attribute__((constructor))
-# define DEP_FINI_ATTRIBUTE static __attribute__((destructor))
```
Comment (t8m) [inline providers/fips/self_test.c:136]: Please drop this extra empty line.

File: providers/fips/self_test.c [hunk 2]
```diff
@@ -103,6 +103,14 @@ BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
...
     return TRUE;
 }
+
+#elif defined(__GNUC__)
+# undef DEP_INIT_ATTRIBUTE
+# undef DEP_FINI_ATTRIBUTE
+# define DEP_INIT_ATTRIBUTE static __attribute__((constructor))
+# define DEP_FINI_ATTRIBUTE static __attribute__((destructor))
+
+
```
Comment (t8m) [inline providers/fips/self_test.c:113]: Please drop this extra empty line

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by jon-oracle** [comment]
Rebased and fixed up review comment.

File: providers/fips/self_test.c (hunk 1 above)
Comment (t8m) [reply providers/fips/self_test.c:135]: You've still kept two empty lines after removal of this block.

**Review by mattcaswell — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@mattcaswell I assume it is still OK.

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for your contribution.

**Pull request closed without merging**